Add logic check only import css from enabled modules

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR based on workarounds @vasilii-b for resolve this remain issue. Default magento import all css from all modules in (app/etc/config.php) no matter it was enabled or not. 
See file less in var/view_preprocessed/pub/static/frontend/Vendor/theme/locale/css
For normal site with small customise module css size is not problem but imagine store have a lot of modules and each module have custom styles. Css output size will be huge and delay browser parse css 

when see file styles-m.less or styles-l.less, you can see this line

```
//@magento_import 'source/_module.less'; // Theme modules
//@magento_import 'source/_widgets.less'; // Theme widgets
```

It's not actually comment code. This is syntax by magento allow import external modules and widgets styles across in all modules and themes

Current exists problem
If we disabled module but we add styles inside theme -> these styles in theme also added to final output instead expect not available in final css.

Benefits/Expects
assume If module disabled so that styles listed below should not be include in output styles
```
@import '../Vendor_Module/css/source/_extend.less'; //Should not include
@import '../Vendor_Module/css/source/_module.less'; //Should not include
@import '../Vendor_Module/css/source/_widgets.less'; //Should not include
```
disabled module is deprecated but some case if we need or know exactly problem come from some module. We can disabled it and also make sure styles come from theme of that module will not parse at final output css.
The benefits => reduce size outputs css file

Some downside/edge-cases

Case 1 (Rarely): If we add styles of disabled module to enabled module. We can't exclude it. I'm not sure in real-world we may got this case (look like it's related with manager code/quality)
Example
Magento_Review -> disabled

in theme luma theme-frontend-luma/Magento_Catalog/web/css/source/_extend.less

```less
@import 'Magento_Review::css/custom.less';
```
theme-frontend-luma/Magento_Review/web/css/custom.less <= styles in here still loaded. Only styles from files, _module.less, widget, _extend will able wipe out

Known limitations
Some variables added/defined in module less. But some other files such as (_extends.less) reference/use/call to it. One case will false if we disabled module
Solutions: Need refactor files to loose tight-coupling


### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#24666

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Magento 2.3.4+
2. blank or luma theme
3. Disable a couple module use less styles: Ex: Magento_Review, Magento_LoginAsCustomerFrontendUi, PageBuilder (note backend function will not work, but frontend styles will expected not available)
4. Add  inside to app/etc/env.php or app/etc/config.php
  ```php
  'static_content_only_enabled_modules' => true,
  ```

5. Run setup upgrade and deploy static content and clear cache
6. Find class belong modules disabled should not available at final css output files(styles-m.css,styles-l.css)

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

mrtuvn committed:
Add logic check only import css from enabled modules

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Unit Tests, Functional Tests EE, Functional Tests CE, Database Compare, Integration Tests

mrtuvn commented:
@magento run Unit Tests

ihor-sviziev commented:
@mrtuvn tbh tho solution looks better and much simpler to me. The only concern I still have about it - it’s backward incompatible and might cause issues for some people/themes.
Maybe we can make it configurable and daabled by default? Later on we'll change it to enabled by default it 2.5

mrtuvn commented:
Hi @engcom-Charlie @engcom-Echo Can you have with unit tests cover such case ?

ihor-sviziev committed:
magento/magento2#32922 Add logic check only import css from enabled modules

Add test coverage

ihor-sviziev commented:
@mrtuvn could you push the following commit / or cherry-pick it to your repository? https://github.com/ihor-sviziev/magento2/commit/3f9feb451a88f4127a36735bdd858955a7f2b22d

mrtuvn commented:
Hi! Thanks you for the tests coverage. I have updated PR branch with your commit

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -108,9 +119,16 @@ protected function replace(array $matchedContent, LocalInterface $asset)
             foreach ($importFiles as $importFile) {
+                $moduleName = $importFile->getModule();
+
+                if ($moduleName &&
+                    !$this->moduleManager->isEnabled($moduleName)) {
+                    continue;
+                }
```
Hi @mrtuvn,
I still have a feeling that this is a backward-incompatible change as it might cause breaking some styles.
As an idea - maybe it's better to introduce some variable in a config.php file that will enable this behavior, similar to `\Magento\Framework\Config\ConfigOptionsListConstants::CONFIG_PATH_FORCE_HTML_MINIFICATION` or `\Magento\Framework\Config\ConfigOptionsListConstants::CONFIG_PATH_SCD_ON_DEMAND_IN_PRODUCTION`?

Later on, in the scope of the 2.5 release line, we might make this flag enabled by default.

What do you think?

mrtuvn replied:
i'm not sure because in the case we disabled module. All js, template related related disabled modules will not render output to html. Only downside of this is can possible happen specific use-cases when disable styling from module A. But in style module A contain some code reference to another component from module B. Disable A may cause some broken visual in module B. Since styles in less from magento quite open and not any restrict you can put styles anywhere you want. In that case we have to perform refactor to migrate styles from A to B

mrtuvn replied:
Currently i'm not clearly about new implement as you mentioned but i think we should have new flag or config for such case
Example css or js file flag output true/false

ihor-sviziev replied:
@mrtuvn,
> Since styles in less from magento quite open and not any restrict you can put styles anywhere you want. In that case we have to perform refactor to migrate styles from A to B

Right, but that's also will affect all the 3rt party modules, that's why it's so important to make it configurable. For the Magento core itself, it's much less relevant.

> Currently i'm not clearly about new implement as you mentioned

You should just add a new flag, and check it in the deployment config, similar to https://github.com/magento/magento2/blob/7c6b6365a3c099509d6f6e6c306cb1821910aab0/app/code/Magento/Config/Model/Config/Structure/ElementVisibility/ConcealInProductionWithoutScdOnDemand.php#L58-L61

ihor-sviziev commented:
@mrtuvn, will you be able to finalize this PR?

mrtuvn commented:
Hi ihor 
Current i have no any clue to start with your new suggest at the moment. This task may need more experiences as backend knowledges
If you can continue this i will very apperciated

ihor-sviziev commented:
@mrtuvn,
yes, this is true, that's why I'm suggesting doing this feature enabled with some flag in the config.php or env.php file.

ihor-sviziev committed:
magento/magento2#32922 Add logic check only import css from enabled modules

Make deploying only enabled modules with flag static_content_only_enabled_modules

ihor-sviziev committed:
magento/magento2#32922 Add logic check only import css from enabled modules

Update tests

ihor-sviziev commented:
@mrtuvn, I just added the additional check from the deployment config. The new flag will be named as `static_content_only_enabled_modules`, default value `false`.
If it is enabled - Magento will include the only enabled modules in the results CSS.

@magento run all tests

mrtuvn committed:
Fix code style

mrtuvn committed:
Fix code style

mrtuvn committed:
Fix code style

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Static Tests, Functional Tests B2B, Functional Tests EE

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -57,31 +67,50 @@ class MagentoImport implements PreProcessorInterface
+     * @return void
      */
-    public function process(\Magento\Framework\View\Asset\PreProcessor\Chain $chain)
+    public function process(Chain $chain): void
```
Please revert adding return type for public or protected methods, that's backward incompatible change

ihor-sviziev replied:
```suggestion
    public function process(Chain $chain)
```

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -97,7 +126,7 @@ public function process(\Magento\Framework\View\Asset\PreProcessor\Chain $chain)
      * @return string
      */
-    protected function replace(array $matchedContent, LocalInterface $asset)
+    protected function replace(array $matchedContent, LocalInterface $asset): string
```
Same here

ihor-sviziev replied:
```suggestion
    protected function replace(array $matchedContent, LocalInterface $asset)
```

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -106,41 +135,52 @@ protected function replace(array $matchedContent, LocalInterface $asset)
+     * @return ThemeInterface
      */
-    protected function getTheme(LocalInterface $asset)
+    protected function getTheme(LocalInterface $asset): ThemeInterface
```
And here

ihor-sviziev replied:
```suggestion
    protected function getTheme(LocalInterface $asset)
```

ihor-sviziev commented on lib/internal/Magento/Framework/Css/Test/Unit/PreProcessor/Instruction/MagentoImportTest.php:
```diff
@@ -129,7 +170,7 @@ public function testProcess($originalContent, $foundPath, $resolvedPath, $foundF
     /**
      * @return array
```
```suggestion
     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
     * @return array
```

ihor-sviziev committed:
Add logic check only import css from enabled modules

Fix static tests

ihor-sviziev commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE

ihor-sviziev commented:
@magento run Functional Tests CE, Functional Tests B2B, Functional Tests EE

ihor-sviziev commented:
@magento run Functional Tests B2B

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9224](https://jira.corp.magento.com/browse/ENGCOM-9224) has been created to process this Pull Request

ihor-sviziev commented:
@magento run Functional Tests B2B

ihor-sviziev commented:
@magento run Functional Tests B2B

engcom-Alfa commented:
Hi @mrtuvn and @ihor-sviziev 
Thanks for your contribution and collaboration!

I have tested with before PR fix code and after PR fix code. The code works well, but the automation test job which is failing needs to be fixed before QA passing it.
Could you please take a look at the failure automation job.

Thanks in advance!

ihor-sviziev commented:
@engcom-Alfa, this failure look not related to changes from this PR. Let's try re-run it few times

@magento run Functional Tests B2B

ihor-sviziev commented:
@engcom-Alfa, all tests are passing now

engcom-Alfa commented:
> @engcom-Alfa, all tests are passing now

Thank you so much for all your contribution @ihor-sviziev

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Preconditions:**

1. Magento installation with sample data


**Manual testing scenario:**

1. Install any new module (eg- Braintree payment) and do the necessary configurations in the admin side.

2. Do one any transaction relevant to the module.
 
3. Disable the module; and delete the content from var/view_preprocessed/* pub/static/*

4. Run static content deploy: bin/magento setup:static-content:deploy -f

5. Look at the generated file pub/static/frontend/Magento/blank/en_US/css/styles-m.css


**Before:** :heavy_multiplication_x: _.payment-method-braintree_ was being available in the file content 

![image](https://user-images.githubusercontent.com/51679138/134120718-05364a26-3ca7-42c9-b466-a3b030053031.png)


**After:** :heavy_check_mark:  _.payment-method-braintree_ is not being available in the file content 

![image](https://user-images.githubusercontent.com/51679138/134120798-134ad8d9-5bd6-4e04-94b8-2b552c7691de.png)

It is relevant to module static content deploy fix and it is not impacting any other feature and hence there is no additional regression is required.

ihor-sviziev commented:
As we discussed with @engcom-Alfa in Slack, that behavior is expected the only with `static_content_only_enabled_modules` flag set to 1 (or true), this need to be re-checked, more details: https://github.com/magento/magento2/pull/32922#issuecomment-914361258

engcom-Alfa commented:
Thanks for all the inputs @ihor-sviziev 
There are 3 validations we have to do in order to merge the PR further.

We have to set the flag 'static_content_only_enabled_modules' => 0 or 1 and needs to check the pub\static\frontend\Magento\blank\en_US\styles-m.css file.
1. With flag set to 1 - we should NOT have braintree styles
2. with flag set to 0 - we should HAVE braintree styles
3. with no flag in the env.php file - we should HAVE braintree styles (equal to 2)
In all the cases, the braintree module is disabled

Observation is : In both the cases when we set to 1 and 0 - the braintree styles have not updated to the css file at all.

engcom-Alfa reviewed (commented):
Based on the above comment, moving it to changes requested.

mrtuvn commented:
Does this work as expect ???
flag 0 or no flag should work as before
only flag 1 will not output styles from disabled module (both from module or from theme area)

ihor-sviziev commented:
@mrtuvn, from our chat with @engcom-Alfa: it always (not depending on the flag) does not output styles from the disabled modules.

Could you double-check that?

mrtuvn commented:
Seem we miss the case if store dont have the flag !

mrtuvn committed:
Update MagentoImport.php

Update check import

mrtuvn committed:
Update MagentoImport.php

Update import check

ihor-sviziev commented:
@mrtuvn sorry, now the changes are not clear to me. Could you explain?

mrtuvn commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -106,24 +135,36 @@ protected function replace(array $matchedContent, LocalInterface $asset)
+                if ($moduleName && $deployOnlyEnabled && !$this->moduleManager->isEnabled($moduleName)) {
+                    continue;
+                }
+
```
added 2 case check here
@ihor-sviziev . What do you think? Feel free correct me based on my understand
Case 1: not available flag check -> ignore disabled styles
Case 2: have flag -> same case 1 (ignore disabled styles)
The rest styles from enabled modules import as usual

ihor-sviziev replied:
If the flag isn't available - it should work the same as on 2.4-develop. Otherwise we'll have backward incompatible change

mrtuvn replied:
updated code can you review one again

mrtuvn committed:
Update MagentoImport.php

Update logic check

ihor-sviziev commented:
Hi @mrtuvn,
It looks like your changes didn't change anything. Could you re-verify if the issue actually exists when you're adding a flag to the config.php file?

ihor-sviziev commented:
@mrtuvn,
Please ping me in Slack and we'll discuss this PR together

tuna2smc committed:
Rework update logic import

mrtuvn commented:
@magento run all tests

mrtuvn committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

tuna2smc committed:
Rename method

mrtuvn committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

mrtuvn commented:
@magento run all tests

ihor-sviziev committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

ihor-sviziev commented:
@magento run all tests

mrtuvn committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

mrtuvn committed:
Update MagentoImport.php

mrtuvn commented:
@magento run Unit Tests , Static Tests

mrtuvn commented:
@magento run all tests

mrtuvn committed:
update logic check

mrtuvn commented:
@magento run Unit Tests, Static Tests, Integration Tests

mrtuvn commented:
cc: @sinhaparul can we consider add this one for community pick/vote this PR

mrtuvn commented:
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, Magento Health Index, Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE, Semantic Version Checker, WebAPI Tests

mrtuvn commented:
@magento run Integration Tests, Functional Tests CE, Functional Tests B2B, Functional Tests EE

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -108,24 +134,51 @@ protected function replace(array $matchedContent, LocalInterface $asset)
-                $importsContent .= $importFile->getModule()
-                    ? "@import $referenceString'{$importFile->getModule()}::{$resolvedPath}';\n"
-                    : "@import $referenceString'{$matchedFileId}';\n";
+
+                if (!$moduleName) {
+                    $importsContent .= "@import $referenceString'{$matchedFileId}';\n";
+                }
+
+                if ($moduleName) {
+                    if (!$deployOnlyEnabled) {
+                        $importsContent .= "@import $referenceString'{$moduleName}::{$resolvedPath}';\n";
+                    } else {
+                        if (!$this->moduleManager->isEnabled($moduleName)) {
+                            continue;
+                        }
+
+                        $importsContent .= "@import $referenceString'{$moduleName}::{$resolvedPath}';\n";
+
+                    }
+                }
```
Maybe we can simplify to something like this?
```suggestion

                if ($moduleName) {
                    if (!$deployOnlyEnabled || $this->moduleManager->isEnabled($moduleName)) {
                        $importsContent .= "@import $referenceString'{$moduleName}::{$resolvedPath}';\n";
                    }
                } else {
                    $importsContent .= "@import $referenceString'{$matchedFileId}';\n";
                }
```

or even

```suggestion

                if (!$moduleName) {
                    $importsContent .= "@import $referenceString'{$matchedFileId}';\n";
                } elseif (!$deployOnlyEnabled || $this->moduleManager->isEnabled($moduleName)) {
                    $importsContent .= "@import $referenceString'{$moduleName}::{$resolvedPath}';\n";
                }
```

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -108,24 +134,51 @@ protected function replace(array $matchedContent, LocalInterface $asset)
+     *
+     * @return bool
+     */
+    protected function hasEnabledFlagDeployEnabledModules(): bool
```
```suggestion
    private function hasEnabledFlagDeployEnabledModules(): bool
```

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -59,31 +69,47 @@ class MagentoImport implements PreProcessorInterface, ResetAfterRequestInterface
+    /**
+     * @var DeploymentConfig
+     */
+    private $deploymentConfig;
```
```suggestion
    private DeploymentConfig $deploymentConfig;
```

ihor-sviziev commented on lib/internal/Magento/Framework/Css/PreProcessor/Instruction/MagentoImport.php:
```diff
@@ -59,31 +69,47 @@ class MagentoImport implements PreProcessorInterface, ResetAfterRequestInterface
+    /**
+     * @var ModuleManager
+     */
+    private $moduleManager;
```
```suggestion
    private ModuleManager $moduleManager;
```

ihor-sviziev reviewed (changes requested):
I see you did some changes recently. Maybe we can cover them with unit tests, so that we won't have regression there in the future?

mrtuvn committed:
Update MagentoImport.php

ihor-sviziev committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (approved):
✔ Approved.

Failing tests look not related to changes from this PR.

engcom-Charlie commented:
Hi @mrtuvn and @ihor-sviziev, 

Thank you for your contribution!

After going through PR’s comment thread here are the summarised points

**Issue addressed in this PR**:  Classes belongs to disabled modules should not available at final css output files(styles-m.css,styles-l.css)

**Solution given & Expected behaviour**: In this PR, “static_content_only_enabled_modules” this flag is introduced. We have to set the flag 'static_content_only_enabled_modules' => 0 or 1 and needs to check the pub\static\frontend\Magento\blank\en_US\styles-m.css file and here are three scenarios with expected output
1. It always (not depending on the flag) does not output styles from the disabled modules.
2. With flag set to 1 like 'static_content_only_enabled_modules' => 1 - we should NOT have braintree styles
3. With flag set to 0 like 'static_content_only_enabled_modules' => 0 - we should HAVE braintree styles
Note: In all the cases, the braintree module is disabled

Now in order to proceed on testing, I have tried to reproduce the issue on latest 2.4-develop 
**Preconditions:**
	1	Magento installation 

**Manual testing scenario:**
	1	Install any new module (eg- Braintree payment) and do the necessary configurations in the admin side.
	2	Do one any transaction relevant to the module. 
	3	Disable the module; and delete the content from var/view_preprocessed/* pub/static/*
	4	Run static content deploy: bin/magento setup:static-content:deploy -f
	5	Look at the generated file pub/static/frontend/Magento/blank/en_US/css/styles-m.css
After 5th step, the .payment-method-braintree is not being available in the file content. So as per the expected output mentioned in point 1, the css is not getting imported after disabling the module hence the **issue is not reproducible**.

**After 2nd step the styles-m.css is:**
<img width="825" alt="image" src="https://github.com/magento/magento2/assets/51680850/5e0bac1c-faf4-4afa-a825-6919791f7c1f">


**After 5th step the style-m.css is:**
<img width="813" alt="image" src="https://github.com/magento/magento2/assets/51680850/5d1560d0-2f30-4715-80ac-0ba01b1a7b70">

As the issue is not reproducible on 2.4-develop as per above steps, can you please check it once again and let me know if I am missing anything. Moving it in Changes Requested.

Thank you!

engcom-Echo committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

engcom-Echo commented:
@magento run all tests

engcom-Echo commented:
### :heavy_check_mark: QA Passed
**Issue addressed in this PR**: Classes belongs to disabled modules should not available at final css output files(styles-m.css,styles-l.css)

**Solution given & Expected behaviour**: In this PR, “static_content_only_enabled_modules” this flag is introduced. We have to set the flag 'static_content_only_enabled_modules' => 0 or 1 and needs to check the pub\static\frontend\Magento\blank\en_US\styles-m.css file and here are three scenarios with expected output

    It always (not depending on the flag) does not output styles from the disabled modules.
    With flag set to 1 like 'static_content_only_enabled_modules' => 1 - we should NOT have braintree styles
    With flag set to 0 like 'static_content_only_enabled_modules' => 0 - we should HAVE braintree styles
    Note: In all the cases, the braintree module is disabled

Now in order to proceed on testing, I have tried to reproduce the issue on latest 2.4-develop
Preconditions:
1 Magento installation

Manual testing scenario:
1 Install any new module (eg- Braintree payment) and do the necessary configurations in the admin side.
2 Create `app/design/frontend/Magento/blank/Paypal_Braintree/web/css/source/_extend.less ` and add custom css
3 Disable the module; and delete the content from var/view_preprocessed/* pub/static/*
4 Run static content deploy: bin/magento setup:static-content:deploy -f
5 Look at the generated file pub/static/frontend/Magento/blank/en_US/css/styles-m.css


**Before:** :heavy_multiplication_x: 
<img width="1093" alt="Screenshot 2024-04-16 at 3 24 02 PM" src="https://github.com/magento/magento2/assets/51681435/ad9674de-75af-463e-a9ce-b90d4936435a">

**After:** :heavy_check_mark:  
<img width="1138" alt="Screenshot 2024-04-16 at 4 31 10 PM" src="https://github.com/magento/magento2/assets/51681435/24236e36-352e-4476-9c69-fc3ac0d80c89">

Moving it to `Extended Testing`

engcom-Echo commented:
@magento run Functional Tests EE,Functional Tests CE,Functional Tests B2B,WebAPI Tests

engcom-Echo commented:
Functional Tests EE,Functional Tests CE,Functional Tests B2B some failure are known failure and other are different on last two run on same commit.

Known Failure
[ACQE-3807](https://jira.corp.adobe.com/browse/ACQE-3807) 
[ACQE-6450](https://jira.corp.adobe.com/browse/ACQE-6450) 
[ACQE-6018 ](https://jira.corp.adobe.com/browse/ACQE-6018) 
[ACQE-5677 ](https://jira.corp.adobe.com/browse/ACQE-5677 )
[ACQE-6438](https://jira.corp.adobe.com/browse/ACQE-6438)
[ACQE-6454](https://jira.corp.adobe.com/browse/ACQE-6454)
[ACQE-6349](https://jira.corp.adobe.com/browse/ACQE-6349)
[ACQE-6438](https://jira.corp.adobe.com/browse/ACQE-6438)
[ACQE-6331](https://jira.corp.adobe.com/browse/ACQE-6331)

Functional Tests B2B
Run1
<img width="1412" alt="b2b-32922" src="https://github.com/magento/magento2/assets/51681435/1c9ae078-33ab-40d6-bb32-9102e51a5635">

Run2
<img width="1432" alt="b2b2-32922" src="https://github.com/magento/magento2/assets/51681435/e0279186-0015-4322-ab74-4566abc17c9b">

Functional Tests CE
Run1
<img width="1440" alt="ce-32922" src="https://github.com/magento/magento2/assets/51681435/f53452bd-b0f7-44bf-9969-8371c02969c9">


Run2
<img width="1440" alt="ce2-32922" src="https://github.com/magento/magento2/assets/51681435/1d01169e-7c43-4664-aee3-1a0be9f9e9af">


Functional Tests EE
Run1
<img width="1413" alt="ee-32922" src="https://github.com/magento/magento2/assets/51681435/e7ba628f-5856-4ca1-b79e-5fdccde5a792">

Run2
<img width="1440" alt="ee2-32922" src="https://github.com/magento/magento2/assets/51681435/09d4206f-00f2-445e-b290-ead79a3b6958">


Web Api Tests failure is known failure
<img width="1409" alt="Screenshot 2024-04-17 at 1 58 33 PM" src="https://github.com/magento/magento2/assets/51681435/a766ec29-5f31-4aba-b166-7f0fe5b325d1">

Hence moving it to `Merge In Progress`

engcom-Charlie committed:
Merge branch '2.4-develop' into add-logic-check-imports-less-for-enabled-modules

engcom-Charlie commented:
@magento run all tests

hostep commented:
PR isn't marked as merged (probably because of the additional commit after it got accepted), but seems to be merged in this bigger merge: https://github.com/magento/magento2/commit/ada8d1d461204d7e1065627384109df1bfa4dcc9

----

I'm curious how this will perform, we've been using the alternative solution to this problem, [using the following patch](https://github.com/magento/magento2/pull/25183#issuecomment-545508802) since 2021 on all our shops and haven't seen any issues with it so far.

Also, the new `static_content_only_enabled_modules` config setting in the deployment config should be documented, can somebody take care of that?

engcom-Charlie commented:
Hi @hostep,

Thats right, This PR got merge in https://github.com/magento/magento2/commit/ada8d1d461204d7e1065627384109df1bfa4dcc9.

The PR status is Open as few commits are not part of merged PR at the time of acceptance. I will be adding these remaining commits in next Mainline PR till then requesting for not to close it.

Thank you!

ihor-sviziev commented:
Finally it marked as merged! Thank you!
